Assignment1/numerical_integration.cpp

```
@@ -0,0 +1,63 @@
+#include "numerical_integration.hpp"
+#include <cmath>
+#include <iostream>
+
+double NumericalIntegrator::integrate(const ScalarFunction& func, double a, double b) const {
+    double sum = 0.0;
+    for (auto i = 0; i < weights.size(); ++i) {
+        double x_mapped = 0.5 * (b - a) * samplingLocations[i] + 0.5 * (a + b);
+        double fx = func(x_mapped);
+        double weighted_fx = weights[i] * fx;
+        sum += weights[i] * func(x_mapped);
+    }
+    sum = sum* 0.5 * (b - a);
+    return sum;
+    
+}
+
+// Gauss-Legendre Quadrature Points and Weights
+std::vector<double> gaussLegendrePoints(double n) {
+    if (n == 1) return {0.0};
+    if (n == 2) return {-0.57735026919, 0.57735026919};
+    if (n == 3) return {-0.77459666924, 0.0, 0.77459666924};
+    if (n == 4) return {-0.86113631159, -0.33998104358, 0.33998104358, 0.86113631159};
```

it's better to raise an exception when `n` is greater than 4 or 0. I tried running your code with `n=5`, the integration result was zero

---

Yeah try adding more cases.